runtime/remote_config.go

```diff
@@ -0,0 +1,327 @@
+func (rc *remoteConfig) Refresh() error {
+	var (
+		err  error
+		stat os.FileInfo
+	)
+	rc.mutex.Lock()
+	defer rc.mutex.Unlock()
+	stat, refresh := rc.checkAndReturnStat()
+	if !refresh {
+		return nil
+	}
+	changedMap, err := rc.reloadConfigAndReturnChanged()
+	if stat != nil && err == nil {
+		rc.currStat = stat
+	}
+	rc.execCallbacks(changedMap)
+	return err
+}
+
+// checkAndReturnStat returns file stat and boolean of `whether we should refresh from config file`
+// don't refresh if config file hasn't change since last refresh
+func (rc *remoteConfig) checkAndReturnStat() (os.FileInfo, bool) {
+	stat, err := os.Stat(rc.config.FilePath)
+	if err != nil {
+		rc.logger.Error("Error stat remote config file " + rc.config.FilePath)
+		return nil, true
+	}
+	if rc.currStat != nil && rc.currStat.ModTime().Equal(stat.ModTime()) && rc.currStat.Size() == stat.Size() {
+		return stat, false
+	}
+	return stat, true
+}
+
+// reloadConfigAndReturnChanged atomically stores remote config properties
+// and return keys been added/removed/updated
+func (rc *remoteConfig) reloadConfigAndReturnChanged() (map[string]bool, error) {
+	bytes, err := ioutil.ReadFile(rc.config.FilePath)
+	changedMap := make(map[string]bool)
+	if err != nil {
+		return changedMap, err
+	}
+	currProps := make(RemoteConfigMap)
+	err = jsonparser.ObjectEach(bytes, func(
+		key []byte,
+		value []byte,
+		dataType jsonparser.ValueType,
+		offset int,
+	) error {
+		currProps[string(key)] = &RemoteConfigValue{
+			bytes:    value,
+			dataType: dataType,
+		}
+		return nil
+	})
+	prevProps := rc.loadConfig()
+	if err == nil {
+		changedMap = changedKeys(prevProps, currProps)
+		rc.props.Store(currProps)
+	}
+	return changedMap, err
+}
+
+// Close wait for all running refresh to complete, then shuts down the client
+func (rc *remoteConfig) Close() {
+	rc.poller.Stop()
+	rc.close <- struct{}{}
+	rc.wg.Wait()
+}
+
+// Subscribe adds a callback function to be executed when config refresh is complete
+// Subscribe takes a string identifier and a function pointer to be called after
+// a completed config refresh
+func (rc *remoteConfig) Subscribe(identifier, key string, fn *func()) {
+	rc.mutex.Lock()
+	defer rc.mutex.Unlock()
+	rc.subscribers[identifier] = &Subscriber{
+		Callback: fn,
+		Key:      key,
+	}
+}
+
+// Unsubscribe will stop an identifer's callback function from being executed upon
+// config refresh
+func (rc *remoteConfig) Unsubscribe(identifier string) {
+	rc.mutex.Lock()
+	defer rc.mutex.Unlock()
+	delete(rc.subscribers, identifier)
+}
+
+func (rc *remoteConfig) execCallbacks(changedMap map[string]bool) {
+	for _, s := range rc.subscribers {
+		if _, ok := changedMap[s.Key]; ok {
+			go (*s.Callback)()
+		}
+	}
+}
+
+func (rc *remoteConfig) poll() {
+	for {
+		select {
+		case <-rc.poller.C:
+			result := "ok"
+			if err := rc.Refresh(); err != nil {
+				result = "err"
+			}
+			rc.scope.Tagged(map[string]string{"result": result}).Counter("remote_config.polling")
```

Comment:
Let's pre-allocate two tagged scopes on the `rc` instance; one for `ok` and one for `err`.

We can also pre-allocate the counter.

Also we need to call `Inc(1)` ;)